### ISSUE TITLE
Database shows stale display name even after it has been changed

### ISSUE THREAD

**Issue by jingyi-zhao-01** [apache/superset#27266]
### Bug description

- when you initially create a database connection, and and change the display name, it still shows the previous display name inside dataset view 

### How to reproduce the bug

- connect your Postgres to superset 
- change the display name
- go to hit the button + DataSet
- in the drop down choose your database 
- you will see the dropdown label shows the stale display name

![Screenshot_20240227_032221](https://github.com/apache/superset/assets/25937657/d5b78308-7ef3-4d4f-9d59-bfd6c70d03b0)
![Screenshot_20240227_032233](https://github.com/apache/superset/assets/25937657/7ab96f3e-cec6-4cd7-8149-0f1a09a3e550)

![Screenshot_20240227_032230](https://github.com/apache/superset/assets/25937657/4362519c-3439-41f2-94ed-8895bef5ab77)

### Screenshots/recordings

![Screenshot_20240227_032206](https://github.com/apache/superset/assets/25937657/cf0f757c-97ab-46dd-b6a1-79723010158f)

### Superset version

master / latest-dev

### Python version

3.9

### Node version

16

### Browser

Chrome

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by jzhao62** [comment]
please assign this issue to me and i will work on it, it is a minor bug though

**Comment by michael-s-molina** [comment]
@jzhao62 I assigned the issue for you. Please link this issue in the fix PR so that it automatically closes when the PR is merged.

**Comment by michael-s-molina** [comment]
Hi @jzhao62. I was not able to reproduce your issue on master following your test instructions. Is there any step missing?

https://github.com/apache/superset/assets/70410625/a88f220c-33e5-4351-b7bd-c17c51315b23

**Comment by jzhao62** [comment]
@michael-s-molina 
yes, you have to create a new database connection and modify its name. ( I think the defaultDB name gets stored in localStrage and being used in other places, this problem might occur for other types of connection as well)

- i have attached a video
1. Make sure you clear localStorage first 
2. the bug can be seen after the 00:01:18 after the video
https://github.com/apache/superset/assets/25937657/32dc3a38-6d33-4daa-9809-d96342a6d5ba

**Comment by rusackas** [comment]
Added a new reviewer to the PR. Hopefully we can get this one across the finish line. Thanks for your patience, and we appreciate the help!

**Comment by rusackas** [comment]
Looks like the PR addressing this issue hasn't been given any love in quite a long time, nor has this issue. We'll probably close both as stale before too long if the PR comments aren't addressed.
